src/main.cpp

```diff
@@ -186,100 +208,192 @@ void addPackOptions(po::options_description& desc)
-            args->verbose = true;
-        }
+    // --gbdxm-file
+    DG_CHECK(vm.count("gbdxm-file") > 0, "No GBDX file specified.");
+    args->gbdxFile = vm["gbdxm-file"].as<string>();
 
-        // --gbdxm-file
-        if(!vm.count("gbdxm-file")) {
-            cerr << "No GBDX file specified." << endl;
-            exit(1);
-        }
+    return args;
+}
 
-        args->gbdxFile = vm["gbdxm-file"].as<string>();
-    }
-    catch(const exception& e) {
-        cerr << "Error reading the arguments: " << e.what() << endl;
-        exit(1);
-    }
+unique_ptr<GbdxmArgs> readShowArgs(const po::variables_map& vm)
+{
+    unique_ptr<GbdxmArgs> args(new GbdxmArgs);
+    args->action = Action::SHOW;
 
-    return args.release();
+    return args;
 }
 
-GbdxmArgs* readShowArgs(const po::variables_map& vm)
+void readFrameworkPackItems(map<string, string>& items,
+                            map<string, string>& options,
+                            vector<string> missingArgs,
+                            const classification::ItemDescriptions& descriptions,
+                            const string& prefix)
 {
-    unique_ptr<GbdxmArgs> args(new GbdxmArgs);
-    args->action = SHOW;
+    for(const auto& item : descriptions) {
+        auto it = options.find(item.name);
+        if(it != options.end()) {
+            items.insert(*it);
+            options.erase(it);
+        } else if(!item.optional) {
+            missingArgs.push_back(prefix + item.name);
+        }
+    }
+}
+
+vector<string> readFrameworkPackArgs(const po::variables_map& vm, GbdxmPackArgs& args)
+{
+    auto type = args.package->type();
+    map<string, string> cliOptions;
+    if(vm.count(type) > 0) {
+        auto argList = vm[args.package->type()].as<vector<string>>();
+        DG_CHECK(argList.size() % 2 == 0, "--%s must have an even number of arguments", type);
+
+        for(auto it = argList.begin(); it != argList.end(); ++it) {
+            auto key = *it++;
+            auto value = *it;
+            cliOptions[key] = value;
+        }
+    }
+
+    auto prefix = string(type) + "-";
+    vector<string> missingArgs;
+
+    readFrameworkPackItems(args.modelFiles, cliOptions, missingArgs, args.identifier->itemDescriptions(), prefix);
 
-    return args.release();
+    map<string, string> options;
+    readFrameworkPackItems(options, cliOptions, missingArgs, args.identifier->optionDescriptions(), prefix);
+    args.package->metadata().setOptions(options);
+
+    DG_CHECK(cliOptions.empty(), "Invalid %s options: %s",
+            args.identifier->type(), join(keys(cliOptions), ", ").c_str());
+
+    return missingArgs;
 }
 
-GbdxmArgs* readPackArgs(const po::variables_map& vm)
+unique_ptr<GbdxmArgs> readPackArgs(const po::variables_map& vm)
 {
-    unique_ptr<GbdxmPackArgs> args(new GbdxmPackArgs);
-    args->action = PACK;
+    auto args = make_unique<GbdxmPackArgs>();
+    args->action = Action::PACK;
 
     // --json
     vector<string> missingFields;
```

Comment:
So it is.   Then:
`missingFields` is reinitialized if both `--json` and `--type` are supplied.  
```
./gbdxm --json static-settings.json --type caffe
```
This is a less obvious use case in gbdxm (in OSN, it's common, because the json can easily be user supplied attributes, but it's still a gotcha).